add second run / test for incremental models in gh action

jeff-dude opened the pull request:
@dot2dotseurat throwing this out there as an idea for now

idea: add to gh action to run & test incremental models twice to ensure all is working still

i'm not sure the commands will work as-is here (using ```,``` without space to filter the run twice)

dot2dotseurat commented:
This is a great idea. I’m afk so I can’t look closely atm but do you want
to test it by modifying an incremental model to trigger the test? Then we
can remove that test commit before merging.

On Thu, Oct 13, 2022 at 6:58 PM j__dude ***@***.***> wrote:

> @dot2dotseurat <https://github.com/dot2dotseurat> throwing this out there
> as an idea for now
>
> idea: add to gh action to run & test incremental models twice to ensure
> all is working still
>
> i'm not sure the commands will work as-is here (using , without space to
> filter the run twice)
> ------------------------------
> You can view, comment on, or merge this pull request online at:
>
>   https://github.com/duneanalytics/spellbook/pull/1771
> Commit Summary
>
>    - 440125d
>    <https://github.com/duneanalytics/spellbook/pull/1771/commits/440125d0ad71b1160b9e0634ec75636506c18341>
>    add second run / test for incremental models
>
> File Changes
>
> (1 file <https://github.com/duneanalytics/spellbook/pull/1771/files>)
>
>    - *M* .github/workflows/dbt_slim_ci.yml
>    <https://github.com/duneanalytics/spellbook/pull/1771/files#diff-196f1549e62fe0f2a83c873307f64d5d3f82f5bee48e309b02b4dab49adb02ac>
>    (6)
>
> Patch Links:
>
>    - https://github.com/duneanalytics/spellbook/pull/1771.patch
>    - https://github.com/duneanalytics/spellbook/pull/1771.diff
>
> —
> Reply to this email directly, view it on GitHub
> <https://github.com/duneanalytics/spellbook/pull/1771>, or unsubscribe
> <https://github.com/notifications/unsubscribe-auth/ACIIUPVAP6KJPLJFQK6746TWDCH2XANCNFSM6AAAAAAREXHD4U>
> .
> You are receiving this because you were mentioned.Message ID:
> ***@***.***>
>

dot2dotseurat commented on .github/workflows/dbt_slim_ci.yml:
```diff
@@ -59,7 +59,7 @@ jobs:
 
       - name: dbt run
-        run: "dbt run --select state:modified,config.materialized:incremental"
+        run: "dbt run --select state:modified,config.materialized:incremental --state ."
```
there's ways to skip this step if it's irrelevant by adding an if statement. https://docs.github.com/en/actions/learn-github-actions/expressions 

you could do something like dbt lst --select config.materialized:incremental in a previous step and set an github arg to false if the output is none. 

tbh we should implement more if statements in this action in general to avoid running steps we don't need. not a bad idea to start here if we want.

jeff-dude replied:
just learning the hard way you either need it or now that you can skip it 😅

jeff-dude replied:
adding the ```--state``` flag worked, but happy to dig into that link tomorrow

dot2dotseurat replied:
This is an example where we set an env variable if thats helpful.     
      - name: Set environment variables
        run: |
          echo "GIT_SHA=$(echo ${{ github.sha }} | tr - _ | cut -c1-8)" >> $GITHUB_ENV
          
         
Something like this could work. `wc -l` will give you a line count. And you could say if the line count is zero to skip the step. 
          ` dbt ls {your selection logic} | wc -l ` 

      - name: Count incremental models
        run: |
          echo "INC_MODEL_COUNT=$(echo dbt ls --select state:modified,config.materialized:incremental --state .  | wc -l)" >> $GITHUB_ENV

dot2dotseurat replied:
> adding the `--state` flag worked, but happy to dig into that link tomorrow

This works but what I'm recommending would make it faster because we wouldn't run a dbt command at all. But it's not super important.

jeff-dude replied:
without looking closely yet, i think we want to run the dbt command though -- the intention of this addition is to make incremental models run & test twice. unless i'm misunderstanding which piece you mention to not run

jeff-dude replied:
without looking closely yet, i think we want to run the dbt command though -- the intention of this addition is to make incremental models run & test twice. unless i'm misunderstanding which piece you mention to not run

dot2dotseurat commented:
I mean, if there are no incremental models, we don’t need to run these
steps at all and it’s adds (a small amount) runtime to the action run these
steps if we don’t need to.

I am just talking about the time it takes for dbt to check the manifests
and see there’s nothing to run.

On Wed, Oct 26, 2022 at 5:51 PM j__dude ***@***.***> wrote:

> ***@***.**** commented on this pull request.
> ------------------------------
>
> In .github/workflows/dbt_slim_ci.yml
> <https://github.com/duneanalytics/spellbook/pull/1771#discussion_r1006218895>
> :
>
> > @@ -59,7 +59,7 @@ jobs:
>          run: "dbt test --select state:new state:modified --state ."
>
>        - name: dbt run
> -        run: "dbt run --select state:modified,config.materialized:incremental"
> +        run: "dbt run --select state:modified,config.materialized:incremental --state ."
>
> without looking closely yet, i think we want to run the dbt command though
> -- the intention of this addition is to make incremental models run & test
> twice. unless i'm misunderstanding which piece you mention to not run
>
> —
> Reply to this email directly, view it on GitHub
> <https://github.com/duneanalytics/spellbook/pull/1771#discussion_r1006218895>,
> or unsubscribe
> <https://github.com/notifications/unsubscribe-auth/ACIIUPTJIX5G3XYLZZUELNDWFGRW7ANCNFSM6AAAAAAREXHD4U>
> .
> You are receiving this because you were mentioned.Message ID:
> ***@***.***>
>

jeff-dude commented:
> I mean, if there are no incremental models, we don’t need to run these steps at all and it’s adds (a small amount) runtime to the action run these steps if we don’t need to. I am just talking about the time it takes for dbt to check the manifests and see there’s nothing to run.
> […](#)
> On Wed, Oct 26, 2022 at 5:51 PM j__dude ***@***.***> wrote: ***@***.**** commented on this pull request. ------------------------------ In .github/workflows/dbt_slim_ci.yml <[#1771 (comment)](https://github.com/duneanalytics/spellbook/pull/1771#discussion_r1006218895)> : > @@ -59,7 +59,7 @@ jobs: run: "dbt test --select state:new state:modified --state ." - name: dbt run - run: "dbt run --select state:modified,config.materialized:incremental" + run: "dbt run --select state:modified,config.materialized:incremental --state ." without looking closely yet, i think we want to run the dbt command though -- the intention of this addition is to make incremental models run & test twice. unless i'm misunderstanding which piece you mention to not run — Reply to this email directly, view it on GitHub <[#1771 (comment)](https://github.com/duneanalytics/spellbook/pull/1771#discussion_r1006218895)>, or unsubscribe <https://github.com/notifications/unsubscribe-auth/ACIIUPTJIX5G3XYLZZUELNDWFGRW7ANCNFSM6AAAAAAREXHD4U> . You are receiving this because you were mentioned.Message ID: ***@***.***>

noted, will look into all this closer 🙌

dot2dotseurat reviewed (approved):
lgtm.

jeff-dude commented:
fyi @soispoke & @augustog
